lib/ansible/modules/crypto/openssl_csr.py

```diff
@@ -247,31 +295,97 @@ def generate(self, module):
                 csr_file.close()
             except (IOError, OSError) as exc:
                 raise CertificateSigningRequestError(exc)
-        else:
-            self.changed = False
+
+            self.changed = True
 
         file_args = module.load_file_common_arguments(module.params)
         if module.set_fs_attributes_if_different(file_args, False):
             self.changed = True
 
-    def remove(self):
-        '''Remove the Certificate Signing Request.'''
+    def check(self, module, perms_required=True):
+        """Ensure the resource is in its desired state."""
+        state_and_perms = super(CertificateSigningRequest, self).check(module, perms_required)
 
-        try:
-            os.remove(self.path)
-        except OSError as exc:
-            if exc.errno != errno.ENOENT:
-                raise CertificateSigningRequestError(exc)
+        self.privatekey = crypto_utils.load_privatekey(self.privatekey_path, self.privatekey_passphrase)
+
+        def _check_subject(csr):
+            subject = csr.get_subject()
+            for (key, value) in self.subject.items():
+                if getattr(subject, key, None) != value:
+                    return False
+
+            return True
+
+        def _check_subjectAltName(extensions):
+            altnames_ext = next((ext.__str__() for ext in extensions if ext.get_short_name() == b'subjectAltName'), '')
+            altnames = [altname.strip() for altname in altnames_ext.split(',')]
+            # apperently openssl returns 'IP address' not 'IP' as specifier when converting the subjectAltName to string
+            # although it won't accept this specifier when generating the CSR. (https://github.com/openssl/openssl/issues/4004)
+            altnames = [name if not name.startswith('IP Address:') else "IP:" + name.split(':', 1)[1] for name in altnames]
+            if self.subjectAltName:
+                if set(altnames) != set(self.subjectAltName):
+                    return False
             else:
-                self.changed = False
+                if altnames:
+                    return False
+
+            return True
+
+        def _check_keyUsage_(extensions, extName, expected, long):
```

Comment:
This check doesn't work as-is and raises issues when running the following playbook two times in a row:

```
---
- hosts: localhost
  tasks:
    - openssl_privatekey:
        path: /tmp/private.key

    - openssl_csr:
        path: /tmp/csr.csr
        privatekey_path: /tmp/private.key                                                                                                                                                      
        commonName: www.ansible.com
```
This is due to the fact thatthe current code relies on `expected` being an array when it could be actually None, hence raising:

```
TypeError: 'NoneType' object is not iterable
```

Here is an alternative implentation that did what was expected, feel free to modify adapt/modify/get ideas from it:

```
usages_ext = [str(ext) for ext in extensions if ext.get_short_name() == extName]

if (not usages_ext and expected) or (usages_ext and not expected):
    return False
elif not usages_ext and not expected:
    return True
else:
    current = [usage.strip() for usage in usages_ext[0].split(',')]
    expected = [long[usage] for usage in expected]
    return current == expected
```